Uploading images to s3

DoctorDib opened the pull request:
Changes:
 - Added Image cropping 
 - Added S3 Upload
 - Added S3 image usage
 - Fixed creating thread uploading to Db

DoctorDib committed:
Adding Image Crop

Changes:
 - Adding image crop component + package to image uploader component
 - In the middle of adding S3 upload

DoctorDib committed:
Creating thread + S3 uploads

Changes:
 - Added image upload to S3 (organised)
 - Added thread creation
 - Removed uneeded (set_vehicle) Db healper function
 - Added image cropping
 - Updated image container (can now crop / remove / make default)

DoctorDib committed:
ESLint error fixes

Changes
 - Fixed all ESLint errors

DoctorDib committed:
Removing ESLint warnings + new errors

Changes:
 - Removed new ESLint warnings
 - Removed new ESLint errors

Note: Still warnings, but most of them are ones that I have no clue about

DoctorDib committed:
Merge branch 'master' into uploading-images-to-s3

BenRGH commented on client/src/Common/Helpers/DB_Helpers.tsx:
```diff
@@ -101,26 +117,30 @@ export const addNewVehicle = async (ownerID: number, data: IInputFields): Promis
-            make: cleanData.make,
-            model: cleanData.model,
-            category: cleanData.category,
+            number_plate: data.number_plate,
```
if you deconstruct data at the top of this func then you don't have to define each property's value in this obj

DoctorDib replied:
Problem is, that will be just as long of a list, I don't think there will be any point?

BenRGH replied:
much tidier, worth it for minimal effort

DoctorDib replied:
I feel like it would be a lot of effort

BenRGH replied:
I'll do it now if you want

DoctorDib replied:
Sure

BenRGH commented on client/src/Common/Helpers/helper.tsx:
```diff
@@ -1,29 +1,53 @@
+    const promises = [];
+
+    const imageArrayLength = imageArray.length;
+    for (let index = 0; index < imageArrayLength; index++) {
```
this could just be a foreach

BenRGH commented on client/src/Common/Helpers/helper.tsx:
```diff
@@ -1,29 +1,53 @@
+
+    return Promise.all(promises).then(() => {
+        console.log('Done');
+        return true;
```
You've defined the return for this func as a promise so I think you should have a resolve here and a reject in the catch

DoctorDib replied:
I don't think we'll be able to since it's already retuning would it should be :Promise<boolean>

BenRGH commented on client/src/Components/ImageCropper/index.tsx:
```diff
@@ -0,0 +1,159 @@
+                    src={imageSrc}
+                    onImageLoaded={onLoad}
+                    crop={crop}
+                    onChange={(c) => setCrop(c)}
```
You've caught these in your recent changes right?

DoctorDib replied:
Changes have been made in a newer PR

BenRGH commented on client/src/Components/ImageCropper/index.tsx:
```diff
@@ -0,0 +1,159 @@
+                    onClick={() => generateDownload()}
+                    className={classes.button}
+                >
+                    {' '}
```
Did we resolve these?

DoctorDib replied:
Changes have been made in a newer PR

BenRGH commented on client/src/Components/ImageUploader/index.tsx:
```diff
@@ -1,32 +1,213 @@
+    setImages: (x:Array<IImageSettings>) => void,
+}
+
+const fileToBase64 = (file:File):Promise<string | ArrayBuffer | null> => new Promise((resolve) => {
```
This should probably be in the helper file

BenRGH commented on client/src/Components/ImageUploader/index.tsx:
```diff
@@ -1,32 +1,213 @@
+    const [cropDialog, setCropDialog] = useState<boolean>(false);
+    const [croppingIndex, setCroppingIndex] = useState<number>(-1);
+
+    const [crop, setCrop] = useState<ICropSettings>(defaultCropSettings);
```
If you're gonna have lots of states it'd probably look nicer all in one group rather than spaced out

DoctorDib replied:
They all have a separate purpose, I'd personally rather keep them in their own states

BenRGH replied:
No I mean remove that line space, it has no purpose

BenRGH commented on client/src/Components/ImageUploader/index.tsx:
```diff
@@ -1,32 +1,213 @@
+        setCroppingIndex(-1);
+    };
+
+    const moveItemInArray = (array:Array<IImageSettings>, from:number, to:number) => {
```
This would be useful in the helper too

BenRGH commented on client/src/Components/ImageUploader/index.tsx:
```diff
@@ -1,32 +1,213 @@
+    };
+
+    const moveItemInArray = (array:Array<IImageSettings>, from:number, to:number) => {
+        array.splice(to, 0, array.splice(from, 1)[0]);
```
It shouldn't modify the original arg array, set the arg type as ReadonlyArray<IImageSettings> and make a new array inside this func with the value of the arg

BenRGH commented on client/src/Components/ImageUploader/index.tsx:
```diff
@@ -1,32 +1,213 @@
+        setImages(newImages);
+        handleClose();
+    };
+    const mapImages = images.map((image: IImageSettings) => (
```
mappedImages is better for this since it's a variable not a func

DoctorDib replied:
Ignore this has been fixed in a later PR

BenRGH commented on client/src/Components/VehicleUploadFields/index.tsx:
```diff
@@ -31,6 +32,8 @@ interface IToolTipMessage {
 const VehicleUploadInputs: React.FC<IImageUploaderProps> = () => {
     const classes: IClasses = styles();
 
+    const [images, setImages] = useState<Array<IImageSettings>>([]);
```
Another one to squish together, since they're all state

DoctorDib replied:
I'm confused this this one?

BenRGH replied:
unnecessary line space again

BenRGH commented on client/src/Components/VehicleUploadFields/index.tsx:
```diff
@@ -217,7 +223,7 @@ const VehicleUploadInputs: React.FC<IImageUploaderProps> = () => {
     );
 
-    const inputComponents = Object.keys(inputFields).map((key: string): any => {
+    const inputComponents = Object.keys(inputFields).map((key: string):React.ReactNode => {
```
If this React.ReactNode works can we change out all the error-inducing FC returns to it please

BenRGH commented on client/package.json:
```diff
@@ -20,17 +21,20 @@
+    "@material-ui/lab": "^4.0.0-alpha.56",
     "@material-ui/pickers": "^3.2.10",
     "@material-ui/styles": "^4.10.0",
     "@types/classnames": "^2.2.10",
```
We should probably move all these type dependencies into the dev dependencies for better build times

DoctorDib committed:
Requested changes

Changes:
 - Updated package.json (now clean)
 - Updated from React.FC to React.ReactElement to remove unwanted ESLint errors.
 - Moved some over to helper
 - Minor changes as requested in PR comments

DoctorDib committed:
Update index.tsx

Changes:
 - Small itty bitty change

BenRGH commented on client/src/Common/Interfaces/interfaces.tsx:
```diff
@@ -67,3 +68,34 @@ export interface IInputFields {
+    y: number,
+}
+
+export interface IParams {
```
Could you not think of slightly more specific name for this exported interface?

DoctorDib replied:
Not really xD - Change it when you do the other comment if you want.

BenRGH committed:
pr fixes
